### ISSUE TITLE
Debug assertion fires in `KeyStats::potential_overlap`

### ISSUE THREAD

**Issue by crepererum** [influxdata/influxdb_iox#2408]
When running IOx locally as a debug build and ingesting a bunch of data via Kafka, I was able to trigger the following assertion:

```text
ERROR panic_logging: thread 'tokio-runtime-worker' panicked at 'assertion failed: self.key_summaries.windows(2).all(|s| s[0].name <= s[1].name)', /home/mneumann/src/influxdb_iox/query/src/provider/overlap.rs:183:9
```

Commit was `4c3056dd914fbfe0bea4b3070e0bde43e6b00351`, so the code is this:

https://github.com/influxdata/influxdb_iox/blob/4c3056dd914fbfe0bea4b3070e0bde43e6b00351/query/src/provider/overlap.rs#L180-L192

Traceback:

<details>

```text
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:515:5
   1: core::panicking::panic_fmt
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/panicking.rs:92:14
   2: core::panicking::panic
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/panicking.rs:50:5
   3: query::provider::overlap::KeyStats<C>::potential_overlap
             at ./query/src/provider/overlap.rs:183:9
   4: query::provider::overlap::group_potential_duplicates
             at ./query/src/provider/overlap.rs:65:20
   5: query::provider::Deduplicater<C>::split_overlapped_chunks
             at ./query/src/provider.rs:494:22
   6: query::provider::Deduplicater<C>::build_scan_plan
             at ./query/src/provider.rs:408:9
   7: <query::provider::ChunkTableProvider<C> as datafusion::datasource::datasource::TableProvider>::scan
             at ./query/src/provider.rs:259:20
   8: datafusion::physical_plan::planner::DefaultPhysicalPlanner::create_initial_plan
             at /home/mneumann/.cargo/git/checkouts/arrow-datafusion-71ae82d9dec9a01c/bd49b86/datafusion/src/physical_plan/planner.rs:311:17
   9: datafusion::physical_plan::planner::DefaultPhysicalPlanner::create_initial_plan::{{closure}}
             at /home/mneumann/.cargo/git/checkouts/arrow-datafusion-71ae82d9dec9a01c/bd49b86/datafusion/src/physical_plan/planner.rs:759:39
  10: core::iter::adapters::map::map_try_fold::{{closure}}
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/adapters/map.rs:89:28
  11: core::iter::traits::iterator::Iterator::try_fold
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/traits/iterator.rs:1972:21
  12: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::try_fold
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/adapters/map.rs:115:9
  13: <core::iter::adapters::ResultShunt<I,E> as core::iter::traits::iterator::Iterator>::try_fold
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/adapters/mod.rs:174:9
  14: core::iter::traits::iterator::Iterator::find
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/traits/iterator.rs:2324:9
  15: <core::iter::adapters::ResultShunt<I,E> as core::iter::traits::iterator::Iterator>::next
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/adapters/mod.rs:156:9
  16: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/alloc/src/vec/spec_from_iter_nested.rs:23:32
  17: alloc::vec::source_iter_marker::<impl alloc::vec::spec_from_iter::SpecFromIter<T,I> for alloc::vec::Vec<T>>::from_iter
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/alloc/src/vec/source_iter_marker.rs:41:20
  18: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/alloc/src/vec/mod.rs:2449:9
  19: core::iter::traits::iterator::Iterator::collect
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/traits/iterator.rs:1748:9
  20: <core::result::Result<V,E> as core::iter::traits::collect::FromIterator<core::result::Result<A,E>>>::from_iter::{{closure}}
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/result.rs:1625:53
  21: core::iter::adapters::process_results
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/adapters/mod.rs:145:17
  22: <core::result::Result<V,E> as core::iter::traits::collect::FromIterator<core::result::Result<A,E>>>::from_iter
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/result.rs:1625:9
  23: core::iter::traits::iterator::Iterator::collect
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/traits/iterator.rs:1748:9
  24: datafusion::physical_plan::planner::DefaultPhysicalPlanner::create_initial_plan
             at /home/mneumann/.cargo/git/checkouts/arrow-datafusion-71ae82d9dec9a01c/bd49b86/datafusion/src/physical_plan/planner.rs:756:39
  25: <datafusion::physical_plan::planner::DefaultPhysicalPlanner as datafusion::physical_plan::planner::PhysicalPlanner>::create_physical_plan
             at /home/mneumann/.cargo/git/checkouts/arrow-datafusion-71ae82d9dec9a01c/bd49b86/datafusion/src/physical_plan/planner.rs:249:28
  26: <query::exec::context::IOxQueryPlanner as datafusion::execution::context::QueryPlanner>::create_physical_plan
             at ./query/src/exec/context.rs:65:9
  27: datafusion::execution::context::ExecutionContext::create_physical_plan
             at /home/mneumann/.cargo/git/checkouts/arrow-datafusion-71ae82d9dec9a01c/bd49b86/datafusion/src/execution/context.rs:485:9
  28: query::exec::context::IOxExecutionContext::prepare_plan
             at ./query/src/exec/context.rs:265:29
  29: server::db::lifecycle::persist::persist_chunks::{{closure}}
             at ./server/src/db/lifecycle/persist.rs:94:29
  30: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/future/mod.rs:80:19
  31: <F as futures_core::future::TryFuture>::try_poll
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-0.3.16/src/future.rs:82:9
  32: <tracker::task::future::TrackedFuture<F> as core::future::future::Future>::poll
             at ./tracker/src/task/future.rs:73:20
  33: <server::db::lifecycle::LockableCatalogPartition as lifecycle::LockablePartition>::persist_chunks::{{closure}}
             at ./server/src/db/lifecycle.rs:217:43
  34: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/future/mod.rs:80:19
  35: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/future/future.rs:119:9
  36: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/task/core.rs:161:17
  37: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/loom/std/unsafe_cell.rs:14:9
  38: tokio::runtime::task::core::CoreStage<T>::poll
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/task/core.rs:151:13
  39: tokio::runtime::task::harness::poll_future::{{closure}}
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/task/harness.rs:461:19
  40: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panic.rs:347:9
  41: std::panicking::try::do_call
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:401:40
  42: __rust_try
  43: std::panicking::try
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:365:19
  44: std::panic::catch_unwind
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panic.rs:434:14
  45: tokio::runtime::task::harness::poll_future
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/task/harness.rs:449:18
  46: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/task/harness.rs:98:27
  47: tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/task/harness.rs:53:15
  48: tokio::runtime::task::raw::poll
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/task/raw.rs:113:5
  49: tokio::runtime::task::raw::RawTask::poll
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/task/raw.rs:70:18
  50: tokio::runtime::task::LocalNotified<S>::run
  51: tokio::runtime::thread_pool::worker::Context::run_task::{{closure}}
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/thread_pool/worker.rs:420:21
  52: tokio::coop::with_budget::{{closure}}
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/coop.rs:106:9
  53: std::thread::local::LocalKey<T>::try_with
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/thread/local.rs:400:16
  54: std::thread::local::LocalKey<T>::with
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/thread/local.rs:376:9
  55: tokio::coop::with_budget
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/coop.rs:99:5
  56: tokio::coop::budget
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/coop.rs:76:5
  57: tokio::runtime::thread_pool::worker::Context::run_task
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/thread_pool/worker.rs:397:9
  58: tokio::runtime::thread_pool::worker::Context::run
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/thread_pool/worker.rs:365:24
  59: tokio::runtime::thread_pool::worker::run::{{closure}}
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/thread_pool/worker.rs:350:17
  60: tokio::macros::scoped_tls::ScopedKey<T>::set
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/macros/scoped_tls.rs:61:9
  61: tokio::runtime::thread_pool::worker::run
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/thread_pool/worker.rs:347:5
  62: tokio::runtime::thread_pool::worker::Launch::launch::{{closure}}
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/thread_pool/worker.rs:326:45
  63: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/blocking/task.rs:42:21
  64: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/task/core.rs:161:17
  65: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/loom/std/unsafe_cell.rs:14:9
  66: tokio::runtime::task::core::CoreStage<T>::poll
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/task/core.rs:151:13
  67: tokio::runtime::task::harness::poll_future::{{closure}}
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/task/harness.rs:461:19
  68: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panic.rs:347:9
  69: std::panicking::try::do_call
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:401:40
  70: __rust_try
  71: std::panicking::try
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:365:19
  72: std::panic::catch_unwind
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panic.rs:434:14
  73: tokio::runtime::task::harness::poll_future
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/task/harness.rs:449:18
  74: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/task/harness.rs:98:27
  75: tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/task/harness.rs:53:15
  76: tokio::runtime::task::raw::poll
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/task/raw.rs:113:5
  77: tokio::runtime::task::raw::RawTask::poll
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/task/raw.rs:70:18
  78: tokio::runtime::task::UnownedTask<S>::run
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/task/mod.rs:373:9
  79: tokio::runtime::blocking::pool::Inner::run
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/blocking/pool.rs:265:17
  80: tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}
             at /home/mneumann/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.10.0/src/runtime/blocking/pool.rs:245:17
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

</details>

I've also modified the code to show the actual names that should be sorted but that aren't. That prints `["bucket_id", "data_spec", "env", "host", "hostname", "node_id", "nodename", "org_id", "partition_id", "role", "url", "time"]` so it seems that something is creating chunks where `time` is the last column. It takes a while however to trigger that issue.

**Comment by crepererum** [comment]
Paging @alamb and @NGA-TRAN , maybe you have an idea what's going on here?

**Comment by NGA-TRAN** [comment]
@crepererum I have followed your instructions and got the red ERROR:
`Aug 30 11:33:10.704 ERROR panic_logging: thread 'tokio-runtime-worker' panicked at 'UNKNOWN', /Users/nga/git/influxdb_iox/query/src/provider/overlap.rs:185:9
thread 'tokio-runtime-worker' panicked at '["bucket_id", "data_spec", "env", "host", "hostname", "node_id", "nodename", "org_id", "partition_id", "role", "url", "time"]', /Users/nga/git/influxdb_iox/query/src/provider/overlap.rs:185:9
stack backtrace:`

Now I need to dig more to have a smaller and deterministic repro

**Comment by NGA-TRAN** [comment]
The problem is wrong assumption on debug_assert!
```
        debug_assert!(self
            .key_summaries
            .windows(2)
            .all(|s| s[0].name <= s[1].name), "{:?}", dbg_self);
```

the `key_summaries` in this context is built from the function [`primary_key`](https://github.com/influxdata/influxdb_iox/blob/55d6140099522a3e2f5d5bc47d78b758cdabab08/internal_types/src/schema.rs#L462) which is lexicographically sorted on their key names except that if there is `time` column, it is always the last one.

In this branch test, the `primary_key` is `"bucket_id", "data_spec", "env", "host", "hostname", "node_id", "nodename", "org_id", "partition_id", "role", "url", "time"`. Hence the above `debug_assert!` fails when it reaches the window ("url", "time").

To fix this, I will eliminate `time` from the debug_assert comparison. I will also look into its child function `potential_overlap_impl` to see if the last `time` column is handled correctly there since now we no longer throwing error here.

**Comment by NGA-TRAN** [comment]
I have checked the function `potential_overlap_impl` in which lexicographical sort on column names will help but not a must. Hence the fix in #2445 will fix this issue
